### PR TITLE
#171380754 make multi city request

### PR DESCRIPTION
**Pull request by muchristian**
**What does this PR do?**
For creating a feature of making multiple cities request

**Description of Task to be completed?**

- add src/database/migrations/20200312122638-create-request.js
- add src/database/models/request.js
- add validations for multi city in validations utils
- write multi city logic in createTripRequest controller function

**How should this be manually tested?**

- Clone `ft-make-multi-city-request-171380754` branch
- Open the mentioned branch in cmd, and then write npm install to install packages
- Run the project by writing in cmd npm run dev.
- Open postman and then run signup route to register -> run login route to authenticate
- Set authentication header and try to send post request of trips route `localhost:3000/api/trips/`
- After the mentioned steps, the trip request will be created and returns  success message and data
_`By the way travelFrom, travelTo and TravelDate, each one should be in array form. for more information look on screenshot below.`_

Mock Data

{
    travelFrom: ['Butare', 'Musanze'],
    travelTo: ['Musanze', 'Rusizi'],
    travelDate: ['2020-08-09', '2020-08-16'],
    travelReason: 'business meeting',
    travelType: 'multi-cities',
    accommodation: true
  },
 {
    travelFrom: ['Butare', 'Musanze', 'Kayonza'],
    travelTo: ['Musanze', 'Kayonza', 'Huye'],
    travelDate: ['2020-08-09', '2020-08-16', '2020-08-24'],
    travelReason: 'business meeting',
    travelType: 'multi-cities',
    accommodation: true
  },

**Any background context you want to provide?**
N/A

**What are the relevant pivotal tracker stories?**
[#171380754](https://www.pivotaltracker.com/n/projects/2435482/stories/171380754)

Capture23

Screenshot (2)

### CODE REVIEW

**Review by since9629 — COMMENTED** [review]
@muchristian could you add screenshots in PR description

File: .eslintrc.json [hunk 1]
```diff
@@ -12,6 +12,8 @@
     "no-console": 0,
     "indent":0,
     "import/first":0,
+    "no-mixed-spaces-and-tabs":0,
+    "no-tabs": 0,
```
Comment (since9629) [inline .eslintrc.json]: @muchristian  why did you disable these rules

File: src/utils/customMessages.js [hunk 2]
```diff
@@ -51,7 +49,11 @@ const messages = {
   usernameExistOrEmpty: 'The username you entered already exists or is empty, if you want to continue update your profile, change the new username or unselect it to remain with the old one',
   passwordChangeSuccess: 'Your password has been changed, login with new password',
   changeEmailNotAllowed: 'You cannot change your email, unselect it if you want to continue updating your profile, otherwise you will not be able to continue',
-  tokenInvalid: 'Authentication failed. Token is invalid'
+  tokenInvalid: 'Authentication failed. Token is invalid',
+  oneWayTripRequestFailed: 'trip request request failed',
+  multiCitiesTripRequestCreated: 'Multiple cities trip request successfuly submitted and waiting for approval',
+  oneWayTripRequestCreated: 'trip request successfuly submitted and waiting for approval',
+  dataLengthErr: 'request should not be more than 2'
```
Comment (since9629) [inline src/utils/customMessages.js]: is a user only allowed to create a travel request for just 2 destinations?

Comment (muchristian) [reply src/utils/customMessages.js]: sure, maximum 2

File: src/utils/validations.js [hunk 3]
```diff
@@ -112,19 +112,15 @@ const validateTripRequest = async (req, res, next) => {
...
-  }
+    try {
+      const validationOutput = travelType === 'one-way' ? await validateOneWayTripRequest(tripRequestInfo)
+      : await validateMultiCitiesTripRequest(tripRequestInfo);
+      validationOutput.userId = req.sessionUser.id;
+      req.body = validationOutput;
+      return next();
+    } catch (validationError) {
+       return errorResponse(res, badRequest, validationError.message);
+    }
```
Comment (since9629) [inline src/utils/validations.js]: why did you delete the switch case here, you could have created your own case block, and added your code there, also remember that 'return-trip' will use it too. Using ternary operator is limiting to only 2 options(one-way and multi-city).

**Comment by descholar-ceo** [comment]
How exactly to test your PR locally? I fetched your branch but I don't know exactly to test it. Could you please add more details on your PR how it should be manually tested?

**Comment by descholar-ceo** [comment]
This is nice, and I tried to test your PR locally, unfortinately I met an issue. Could you help me with that please?

File: src/utils/validations.js [hunk 4]
```diff
@@ -112,19 +112,15 @@ const validateTripRequest = async (req, res, next) => {
...
-        req.body = validationOutput;
-        return next();
-      } catch (validationError) {
-        return errorResponse(res, badRequest, validationError.message);
-      }
-    default:
-      return errorResponse(res, badRequest, invalidTravelType);
-  }
+    try {
+      const validationOutput = travelType === 'one-way' ? await validateOneWayTripRequest(tripRequestInfo)
```
Comment (descholar-ceo) [inline src/utils/validations.js]: This way you used here, will be difficult when there will come to more than to `tripTypes`, as @karamuka said, you should have tried tp use switch case

**Review by descholar-ceo — COMMENTED** [review]
Could you help me with this error message? I don't understand the field which I didn't fill
Screenshot from 2020-03-22 23-16-11

File: src/utils/validators.js [hunk 5]
```diff
@@ -50,4 +50,34 @@ static async validateOneWayTripRequest(tripRequestData) {
...
+  const { createValidationErrors } = Validators;
+  const joiArray = Joi.array().items(Joi.string()).required();
+  const schema = Joi.object({
+    travelTo: joiArray
+    .messages(createValidationErrors('string', invalidTravelTo)),
+    travelFrom: joiArray
+    .messages(createValidationErrors('string', invalidTravelFrom)),
+    travelReason: Joi.string().required()
+    .messages(createValidationErrors('string', invalidTravelReason)),
+    travelType: Joi.string()
```
Comment (the22mastermind) [inline src/utils/validators.js]: Could you use an array of pre-defined values, for eg. ["one-way", "return-trip", "multi-city"] as the only valid inputs? Because currently the type can be anything and it doesn't make sense to have a travel type of, let's say "hello"
Screenshot from 2020-03-23 00-45-06

**Review by the22mastermind — COMMENTED** [review]
@muchristian Good job but I have a few questions and suggestions.
1. Is there a reason why you limited the trip request to have only 2 destinations?
Screenshot from 2020-03-23 01-04-45

2. When I send the same data twice, it doesn't get saved in the database but this error pops up and I think you should catch it where you save in the database and return a meaningful message because currently it shows that the data is saved which is not the case.
Screenshot from 2020-03-23 00-49-46

**Review by descholar-ceo — COMMENTED** [review]
You did a good job to improve, and the error messages you tried to change them, that is nice, I like it. However, I have a few comments on your work:

1. When creating a trip, could you return the trip which has been just created to the user? As you see on the screenshot, the data property is null, though this trip is inserted on the DB, so could you fix this issue?
Screenshot from 2020-03-23 22-39-18

2. Secondly, when a user tries to resubmit this request, it still shows that the data has been inserted in the DB, whereas is not, and on the console of the app it throws this error on the screenshot:
Screenshot from 2020-03-23 22-42-50
===>Could you catch this error and you return an appropriate message? easy to be understood by the user?

3. When a user sends `travelFrom` cities in this format `"San Francisco, Seoul"` it accepts it but when it is sent in this format `"San Francisco", "Seoul"` it takes the first city only, it seems like although it is an array but it taking only the forst elementt, it is not taking all of the two elements passed in the array, it is the same thing on the `travelDate` when you enter two travelDates in this format `"2020-04-28", "2020-04-29"` it is taking only the first one, so could you fix this issue too?

4. I recommend you to update your PR to add more details about testing your PR and formats of the request body that we need to send inorder to facilitate us to test your PR.

**Comment by the22mastermind** [comment]
@muchristian Could you please check the validation of travelDate?
1. TravelDate should enforce a consistent format
2. You should then check if the date is not in the past.
Screenshot from 2020-03-24 12-58-29
Screenshot from 2020-03-24 12-59-46
3. Please handle that `SequelizeUniqueConstraintError` **Tip:** Do it in request.service.js

**Review by Ugizwenayo-Divine — CHANGES_REQUESTED** [review]
you did great but I have something, I would like to suggest
1. I tried to request a trip where the first destination was different from the second origin, I got an error that they have to be the same and it's good but after that error message, I sent a correct one and I got the success message but within data, there were also all those requests which failed  

Capturedata


2. I think you need to validate travelFrom and travelTo in such a way a user enters butare and later Butare has to be considered the same

CaptureCapital


3.  I saw that there is no validation when a user submits one travel date

CaptureDate


4. I saw that a user can provide one travelTo and request is accepted

CaptureTo

**Review by the22mastermind — APPROVED** [review]
Well done @muchristian 
Keep it up!

**Review by Ugizwenayo-Divine — COMMENTED** [review]
Well done, you implemented the feedbacks, but also try to look into  those few comments
I've seen that a user can create a trip with empty travelReason
which I think, is not very good

I have seen also that you can create a trip where the first depart place is the same as the final destination. which seems like: I'm in Kigali and I'm requesting a trip to go to Kigali and my stops are Butare, Musanze .

Finally, validate that  when travel to are 3, travel from must be 3 and travel date as well

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
@muchristian You have done great work, and you did well to respond to some feedback you got, that is great, however, I have some feedback that is worthwhile. 
1. I saw that you cannot create a trip like the following : 
    >Washington DC ===> Seoul : 2020-04-20
    >Seoul ===> Tokyo : 2020-04-20
    >Tokyo ===> Yokohama : 2020-04-20
And this should be allowed.

2. And also something we didn't notice since one-way, return trip is that I saw guys we are not validating trip which has travelFrom = travelTo; is that okay? In your case, this trip is allowed and I think it is meaningless:
    >Seoul ===> Seoul 
    >Seoul ===> Seoul
    >Seoul ===> Seoul

File: src/controllers/request.controller.js [hunk 6]
```diff
@@ -4,33 +4,84 @@ import statusCodes from '../utils/statusCodes';
...
+ const { successResponse, errorResponse, } = responseHandlers;
+ let data = [];
+
+
+ /**
+   * @param {Request} arr array
+   * @param {Request} userId authenticated user id
+   * @returns {Object} return error
+   */
+ const errorHandler = async (arr, userId) => {
```
Comment (Muhire-Josue) [inline src/controllers/request.controller.js]: can you move this `errorHandler` to `/src/utils` because I don't think it's a good idea to implement `errorHandlers` in controller, we can call them in controller no problem but implement them in `/src/utils`

File: src/controllers/request.controller.js [hunk 7]
```diff
@@ -4,33 +4,84 @@ import statusCodes from '../utils/statusCodes';
...
-    try {
-      const requestData = req.body;
-      const newTrip = await handleTripRequest(requestData);
-      return successResponse(res, created, oneWayTripRequestCreated, undefined, newTrip);
-    } catch (dbError) {
-      return errorResponse(res, badRequest, duplicateTripRequest);
-    }
-  }
+    const requestData = req.body;
+    const { travelFrom } = req.body;
```
Comment (Muhire-Josue) [inline src/controllers/request.controller.js]: since you did this `const requestData = req.body;` you can do this `const { travelFrom } = requestData ;`

**Review by descholar-ceo — APPROVED** [review]

File: src/controllers/request.controller.js [hunk 8]
```diff
@@ -4,33 +4,84 @@ import statusCodes from '../utils/statusCodes';
...
+    const { travelDate } = requestData;
+    if (travelFrom instanceof Array) {
+      travelFrom.map((item, i) => data.push({
+        ...req.body,
```
Comment (Muhire-Josue) [inline src/controllers/request.controller.js]: same thing here, you can replace `req.body` with `requestData `

File: src/controllers/request.controller.js [hunk 9]
```diff
@@ -4,33 +4,84 @@ import statusCodes from '../utils/statusCodes';
...
+        travelFrom: item,
+        travelTo: req.body.travelTo[i], 
+        travelDate: req.body.travelDate[i],
```
Comment (Muhire-Josue) [inline src/controllers/request.controller.js]: here too, 
```
travelTo: requestData.travelTo[i],
travelTo: requestData .travelDate[i]

```

File: src/controllers/request.controller.js [hunk 10]
```diff
@@ -4,33 +4,84 @@ import statusCodes from '../utils/statusCodes';
...
+  }
+  return error;
+ };
+
+/**
+   * @param {Request} arr array
+   * @returns {Object} return await function
+   * @description insert data in db by looping
+   */
+const addMulti = async (arr) => {
```
Comment (Muhire-Josue) [inline src/controllers/request.controller.js]: Could you please rename this function to something more descriptive? Also, I think you should move this function to `utils` and call it in the controller.

**Review by Muhire-Josue — COMMENTED** [review]
great work

**Review by Muhire-Josue — APPROVED** [review]
LGTM

File: .eslintrc.json [hunk 11]
```diff
@@ -12,6 +12,9 @@
     "no-console": 0,
     "indent":0,
     "import/first":0,
+    "no-restricted-syntax":0,
+    "no-await-in-loop":0,
+    "newline-per-chained-call":0,
```
Comment (since9629) [inline .eslintrc.json]: why disable these rules?

**Review by Ugizwenayo-Divine — COMMENTED** [review]
CaptureThis
would you work on this

CaptureThis2
 
try to handle this error, the app crashes if you send 2 travel date at the time you would have sent 3 travel date

**Review by Ugizwenayo-Divine — APPROVED** [review]
Well done @muchristian 
congratulation

**Review by since9629 — APPROVED** [review]

**Comment by kipropbrian** [comment]
Can you rebase this. @muchristian

**Comment by kipropbrian** [comment]
@muchristian Please  attend all the previous feedback as soon as possible

File: src/services/request.service.js [hunk 12]
```diff
@@ -95,45 +134,110 @@ export default class RequestService {
...
-        travelDate,
-        returnDate,
-        accommodation
-      },
-      { where: { id: requestId } }
-    );
-  }
+      returnDate
+     },
+     { where: { requestId: requestId } }
```
[the remark this answers is not in the document]
Comment (the22mastermind) [reply src/services/request.service.js:231]: { where: { requestId } }

File: tests/data/insert-sample-data-in_db.js [hunk 13]
```diff
@@ -9,8 +9,11 @@ VALUES(6,1,'Welcome to the new world',NOW(),NOW())`;
 export const INSERT_USER = `
 INSERT INTO users ("id","firstName","lastName","email","username","password","gender","address","role","isVerified","createdAt","updatedAt")
 VALUES(15,'updator','updator',updator@gmail.com,'updator','Updator@0000','male','kigali','requester',true,NOW(),NOW())`;
-export const INSERT_REQUEST = `
```
Comment (the22mastermind) [inline tests/data/insert-sample-data-in_db.js:12]: Why did you delete this variable?

Comment (muchristian) [reply tests/data/insert-sample-data-in_db.js:12]: i had to change how queries where written because the erd has changed

File: src/services/request.service.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (muchristian) [reply src/services/request.service.js:231]: i did it

**Comment by the22mastermind** [comment]
@muchristian you need to change the error message in `request.test.js` line 651.
Also use console.log(res.body) in the remaining 5 tests that are failing to get a hint of the response.

**Review by Ugizwenayo-Divine — COMMENTED** [review]
You did great job Chris and also think about those
within code climate file put new line at the end
update is not working on multi-city trip request

File: .codeclimate.yml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+## other configuration excluded from example...
+exclude_patterns:
+- "src/database/"
+- "tests/"
+- "src/services"
```
Comment (descholar-ceo) [inline .codeclimate.yml]: Why did you exclude `services from codeclimate review`?

File: .codeclimate.yml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+## other configuration excluded from example...
+exclude_patterns:
+- "src/database/"
+- "tests/"
+- "src/services"
+- "src/utils/validations.js"
```
Comment (descholar-ceo) [inline .codeclimate.yml]: Could you solve this trailing space?

File: src/middlewares/authentication.js [hunk 16]
```diff
@@ -56,7 +56,7 @@ export default class Authentication {
         req.sessionUser = decodedToken;
         return next();
       });
-    } catch (error) {
+    } catch (error) {
```
Comment (descholar-ceo) [inline src/middlewares/authentication.js:59]: why did you touch on this middleware? was not working?

**Review by descholar-ceo — COMMENTED** [review]
Screenshot from 2020-05-05 11-29-53
I think this message should be made appropriate because a usual would not know what is ISO 8601 date format even I, I don't what it is.

File: src/services/request.service.js [hunk 17]
```diff
@@ -95,45 +138,106 @@ export default class RequestService {
...
       accommodation
+     },
+     { where: { id: requestId } }
+   );
+   /*
+   return sequelize.query(`UPDATE requests SET requests.travelReason = ${travelReason},
+   requests.travelType = ${travelType}, requests.accommodation = ${accommodation},
+   sub_requests.travelFrom = ${travelFrom}, sub_requests.travelTo = ${travelTo},
+   sub_requests.travelDate = ${travelDate}, sub_requests.returnDate = ${returnDate}
+   FROM sub_requests WHERE requests.id = sub_requests.requestId OR requests.id = ${requestId}`);
```
Comment (descholar-ceo) [inline src/services/request.service.js]: could you remove these commented codes?

File: src/services/request.service.js [hunk 18]
```diff
@@ -95,45 +138,106 @@ export default class RequestService {
...
+      returnDate
+     },
+     { where: { requestId } }
+   );
+   /*
+   return sequelize.query(`UPDATE requests SET requests.travelReason = ${travelReason},
+   requests.travelType = ${travelType}, requests.accommodation = ${accommodation},
+   sub_requests.travelFrom = ${travelFrom}, sub_requests.travelTo = ${travelTo},
+   sub_requests.travelDate = ${travelDate}, sub_requests.returnDate = ${returnDate}
+   FROM sub_requests WHERE requests.id = sub_requests.requestId OR requests.id = ${requestId}`);
```
Comment (descholar-ceo) [inline src/services/request.service.js]: same case here, you can remove the commented codes

File: src/services/request.service.js [hunk 19]
```diff
@@ -1,10 +1,26 @@
...
 const { ACCEPTED, } = tripRequestsStatus;
 
+ /**
+   *@description Saves trip request details in database
+   * @param {Object} field trip request data
+   * @returns {Object} Saved trip request details
+    */
+const getModel = (field) => {
+  let val = null;
+  val = request;
```
Comment (descholar-ceo) [inline src/services/request.service.js]: I think you could even do the initialization of this variable on line 14, there was no need to initialize it on the other line

File: src/services/request.service.js [hunk 20]
```diff
@@ -1,10 +1,26 @@
+import { Op } from 'sequelize';
 import models from '../database/models';
 import tripRequestsStatus from '../utils/tripRequestsStatus.util';
 
-const { request, sequelize, Sequelize } = models;
-const { Op } = Sequelize;
+const { request, trips, sequelize, Sequelize } = models;
```
Comment (descholar-ceo) [inline src/services/request.service.js]: Since you did not use this `Sequelize` with a capital letter, there is no need to declare it as a variable, could you remove it?

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
@muchristian you did well however, I have a few concerns about this PR, : 

1. Since the request table has been split into two different tables, trips and requests , this endpoint GET /api/trips to get all tripRequests, it is not giving the expected results, it is returning only requests but no trips, so I was wondering how will we retrieve trips

2. Don't exclude `services` from code climate review please, we need the quality of the codes from services.

3. Your commit message has less info comparing from what you have done on this PR,

4. Take care of all comments I added in your codes.

**Review by Ugizwenayo-Divine — COMMENTED** [review]
would you solve code climate issue and CI issue

**Comment by descholar-ceo** [comment]
> would you solve code climate issue and CI issue

@Ugizwenayo-Divine I am currently working on those issues in this PR [Fix Code climate issues](https://github.com/Stackup-Rwanda/stackup2-barefoot-backend/pull/52)

**Review by the22mastermind — APPROVED** [review]
Well done @muchristian and @de-scholar 
Keep up the good work.

**Review by descholar-ceo — APPROVED** [review]

**Review by Ugizwenayo-Divine — APPROVED** [review]
great job Chris
LGTM

**Comment by kipropbrian** [comment]
- This is the response I get when I create a multi-city trip
```
{
    "message": "trip request successfuly submitted and waiting for approval",
    "data": [
        {
            "id": 3
        },
        {
            "travelFrom": "Butare",
            "travelTo": "Musanze",
            "travelDate": "2020-08-09T00:00:00.000Z",
            "requestId": 3
        },
        {
            "travelFrom": "Musanze",
            "travelTo": "Rusizi",
            "travelDate": "2020-08-16T00:00:00.000Z",
            "requestId": 3
        }
    ]
}
```

Why is `id:3` on its own? It does not make sense. 

Could you stick to the object structure returned while fetching data
```
{
  message: <message>,
  data: <rows>,
  count: <count of data/rows> (only add this property)
}
```

Make sure that the data part looks something like this 
```
{
  "request": { 
                     trips: [{
                               }]
                     }
}
```
Trips should be part of requests. 


- Travel type should be "multi-city" not multi cities.

**Comment by kipropbrian** [comment]
We also need a GET for all Multi-city request trips, or GET all trips should also include mutlicity requests.

File: src/middlewares/ratingChecker.js [hunk 21]
```diff
@@ -12,7 +13,7 @@ const {
     notBookingOwner,
     beforeTripDate, 
 } = customMessages;
-const { findTripRequestById } = RequestService;
+// const { findTripRequestById } = RequestService;
```
Comment (kipropbrian) [inline src/middlewares/ratingChecker.js]: Remove the comments

File: src/middlewares/ratingChecker.js [hunk 22]
```diff
@@ -47,7 +48,8 @@ const isTripRequestValidIsYours = async (req, res, next) => {
  * get request id from request body, checks if the request exist or not
  */
 const didRequestBookAccommodation = async (req, res, next) => {
-    const booking = await getBookedAccommodationByRequestId(req.body.requestId);
+    // const booking = await getBookedAccommodationByRequestId(req.body.requestId);
```
Comment (kipropbrian) [inline src/middlewares/ratingChecker.js]: here too
